### ISSUE TITLE
Add login URL to scope and context

### ISSUE THREAD

**Issue by voelklflo** [zaproxy/zaproxy#2130]
Hi,

I tried recently to do "Spider as User" on one of the products I am developing. I wanted to add ZAProxy CI-Jobs to our Jenkisn using the [Jenkins ZAProxy Plugin](https://wiki.jenkins-ci.org/display/JENKINS/ZAProxy+Plugin).

My application uses a central authentication form. The autentication form is accessed by a url diffrent from the main url of the application.

The problem I faced was, that whatever I tried either the login page was "Out of Scope" or "Out of Context" or likely both.

If I understand the usage of the "Spider as User"-option right, the specified login url could/should always be part o the context and in scope. This would fix my problem and make the usage of the ZAProxy more obvious.

Thank you for all the work on this great product.

Regards,
Florian

**Comment by thc202** [comment]
The login page does not need to be in context (or in scope), the authentication should happen behind the scenes and the spider should not be seeing it.

Is the authentication working as expected without using the Jenkins plugin?

**Comment by voelklflo** [comment]
The problem seems to be that the Spider fails before the login is
attempted/reattempted.

No. When trying to spider without Jenkins plugin ZAP also reports "OUT OF
SCOPE".

The "active scan" works in a trained session though.

2015-12-21 19:17 GMT+01:00 thc202 <notifications@github.com>:

**Comment by thc202** [comment]
OK, could you provide the steps that you are doing to set up the authentication?

**Comment by gorivishal** [comment]
I am stuck in a similar situation. I have included all the URLs in scope in context. Also, marked the context in scope using zap.context.set_context_in_scope(contextname,True). 

I am using ZAP API. Everything works fine while feeding ZAP through proxy. What could be missing in manual configuration?
Steps:
# Create context and set it in scope.
# Include list of URLs necessary in the context. (Regex: URL+".*")
# Exclude \Qlogout\E from spider
# Session Management - Default is cookieBasedSessionManagement
# Set log in indicator
# Enabling session handling.
# Spider everything included in context.

**Comment by thc202** [comment]
Are you also setting up the authentication method? (e.g. [1])
How are you starting the spider? Are you specifying a user?

[1] https://github.com/zaproxy/zaproxy/wiki/FAQformauth

**Comment by gorivishal** [comment]
I think I was doing it write but the session handling in the application I am trying to scan has been configured quite weirdly. However, the "Enable Session Handling" button under "Edit" menu automagically made it work. Is there an API call I can use to check that button for my scan? I really couldn't find one. Authentication is being taken care through Selenium now where I also made Selenium to use a pre-configured firefox profile and proxy it through ZAP. But I cant make ZAP to remember the cookies correctly.

**Comment by gorivishal** [comment]
It would really be great if I can get an answer to this question:
>> Is there an API call I can use to check "Enable Session Handling" button under "Edit" menu  button for my scan?

**Comment by thc202** [comment]
It can be enabled with `core/action/setOptionHttpStateEnabled/?Boolean=true` (it must be done before sending any request through the proxy).

But using HTTP Sessions API is probably better, as it gives you more control on what cookies are tracked and for which websites.

**Comment by gorivishal** [comment]
Awesome! The application I am working on is badly configured to include session tokens in the html body. Manual control is proving difficult to implement. This is very helpful 👍 . Thanks.

**Comment by thc202** [comment]
Closing, this seems to be answered/resolved.
